velox/exec/prefixsort/PrefixSortAlgorithm.h

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/base/SimdUtil.h"
+
+namespace facebook::velox::exec::prefixsort {
+
+namespace detail {
+/// Provides LegacyRandomAccessIterator but not ValueSwappable semantic
+/// iterator for PrefixSort`s sort algorithm, see:
+/// https://en.cppreference.com/w/cpp/algorithm/sort
+/// https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+/// https://en.cppreference.com/w/cpp/named_req/ValueSwappable
+/// The difference is operator*(), the LegacyInputIterator return "reference,
+/// convertible to value_type", see:
+/// https://en.cppreference.com/w/cpp/named_req/InputIterator
+/// In contrast, PrefixSortIterator returns char* point to the prefix data
+/// (not a "value_type" but a memory buffer store normalized keys) for memcmp
+/// and memcpy operators, means not ValueSwappable.
+/// This is also the reason we can not use std::sort directly, and implement
+/// quickSort in PrefixSortContext.
+struct PrefixSortIterator {
+  PrefixSortIterator(char* prefix, const uint64_t entrySize)
+      : prefix_(prefix), entrySize_(entrySize) {}
+
+  PrefixSortIterator(const PrefixSortIterator& other)
```

Comment:
```
explicit PrefixSortIterator(const PrefixSortIterator& other) :  PrefixSortIterator(other.prefix, other.entrySize_)
```

Reply:
Could you detail the purpose to add explicit for a copy construct here ?  @xiaoxmeng

Reply:
@xiaoxmeng 
The problem to use explict here is that we lose copy_initialization , ref:

https://en.cppreference.com/w/cpp/language/copy_initialization

code like this could not work.

```
  FOLLY_ALWAYS_INLINE PrefixSortIterator operator-(const uint64_t i) const {
    PrefixSortIterator result = *this;
    result.prefix_ -= i * entrySize_;
    return result;
  }
```

about the explict copy-construct here`s a doc as ref :

https://quuxplusone.github.io/blog/2023/04/08/most-ctors-should-be-explicit/

>So never mark a single-argument copy or move constructor as explicit. But do continue to mark your zero-argument constructor explicit; there’s no problem with the initialization of a1 here.